_posts/2016-02-04-utilizing-enumerable-and-delegate.markdown

```diff
@@ -0,0 +1,137 @@
+    end
+  end
+
+  def unsorted_newsletters(file_name)
+    file_contents(file_name).select { |newsletter| newsletter =~ /.*\.pdf/i }
+  end
+
+  def file_contents(file_name)
+    Dir.entries("app/assets/newsletters/#{file_name}")
+  end
+end
+
+{% endhighlight %}
+
+Here are my two objects a NewsletterController and a NewsletterFinder. Between these objects there are a lot of dependencies. NewsletterController knows that there is a NewsletterFinder object and it knows two methods that can be called on the NewsletterFinder object. The NewsletterFinder is also dependent on the folder names. I feel there is little to be done regarding the first dependency however we can improve the code to reduce the other dependencies.
+
+Step 1 - passing an argument not calling a method
+In the NewsletterController request a new instance of NewsletterFinder and pass an argument. The argument represents the folder name which holds the documents desired by the view.
+
+This requires the NewsletterFinder object to have an initialize method which accepts an argument. 
+
+This change results in the NewsletterController having less knowledge about the NewsletterFinder class which reduces dependencies. It also removes all responsability of the NewsletterFinder class to know what the folder name is. This means changes can be made to the folder name without needing to change anything within the NewsletterFinder class.
+
+{% highlight ruby %}
+class NewslettersController < ApplicationController
+  def index
+    @flyers = NewsletterFinder.new("monthly_flyers")
+    @news = NewsletterFinder.new("news_documents")
+  end
+end
+
+class NewsletterFinder
+  def initialize(type)
+    @type = type
+  end
+
+  def file_contents
+    Dir.entries("app/assets/newsletters/#{@type}")
+  end
+end
+
+{% endhighlight %}
+
+Step 2 - reducing responsabilities down to one
+The NewsletterFinder has two methods `flyers` and `news`. These methods have two responsabilities, they define a variable which gets passed through the method chain and they remove the .pdf from the file name. 
+
+Fortunately the Step 1 means the variable assignment can be cut out as that is now being passed in from the controller. All that is left is two copies of the same logic which can be made into one method called `newsletter_list`.
+
+{% highlight ruby %}
+def newsletters_list
+  newsletters_with_file_extension.map do |newsletter|
+    newsletter.gsub(/\.pdf/, "")
+  end
+end
+{% endhighlight %}
+
+Step 3 - undefined method error each
+Making these changes results in an error when I refresh the view. Undefined method each for `@flyers.each`. Each is a method which is included in the `Enumerable Module`. The array class includes the `Enumerable Module` and therefore if you call .each on an array you get the expected behaviour. The `NewsletterFinder` object is not an array, nor does it include the `Enumerable Module`. `@flyers` is an instance of `NewsletterFinder` and hence it does not know how to respond to `each`.
```

Comment:
Array's have the enumerable module mixed in but have their own definition of each.